lib/optimizer/tests/Transformer/PreloadHeroImageTest.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace AmpProject\Optimizer\Transformer;
+
+use AmpProject\Dom\Document;
+use AmpProject\Optimizer\Configuration\PreloadHeroImageConfiguration;
+use AmpProject\Optimizer\Error;
+use AmpProject\Optimizer\ErrorCollection;
+use AmpProject\Optimizer\Tests\ErrorComparison;
+use AmpProject\Optimizer\Tests\MarkupComparison;
+use AmpProject\Optimizer\Tests\TestMarkup;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test the PreloadHeroImage transformer.
+ *
+ * @package ampproject/optimizer
+ */
+final class PreloadHeroImageTest extends TestCase
+{
+    use ErrorComparison;
+    use MarkupComparison;
+
+    /**
+     * Provide the data to test the transform() method.
+     *
+     * @return array[] Associative array of data arrays.
+     */
+    public function dataTransform()
+    {
+        $input = static function ($body, $preloads = '') {
+            return TestMarkup::DOCTYPE . '<html ⚡><head>'
+                   . TestMarkup::META_CHARSET . TestMarkup::META_VIEWPORT . $preloads . TestMarkup::SCRIPT_AMPRUNTIME
+                   . TestMarkup::LINK_FAVICON . TestMarkup::LINK_CANONICAL . TestMarkup::STYLE_AMPBOILERPLATE . TestMarkup::NOSCRIPT_AMPBOILERPLATE
+                   . '</head><body>'
+                   . $body
+                   . '</body></html>';
+        };
+
+        $output = static function ($body, $preloads = '') {
+            return TestMarkup::DOCTYPE . '<html ⚡ i-amphtml-layout=""><head>'
+                   . TestMarkup::META_CHARSET . TestMarkup::META_VIEWPORT . $preloads . TestMarkup::SCRIPT_AMPRUNTIME
+                   . TestMarkup::LINK_FAVICON . TestMarkup::LINK_CANONICAL . TestMarkup::STYLE_AMPBOILERPLATE . TestMarkup::NOSCRIPT_AMPBOILERPLATE
+                   . '</head><body>'
+                   . $body
+                   . '</body></html>';
+        };
+
+        return [
```

Comment:
There are two test cases here. Both of them are for error scenarios. Shouldn't there be one where there is no error?

Reply:
In other words, perhaps `throws error when scrset detected on image to be preloaded` should be split into two.

Reply:
Ther spec tests are already quite exhaustive for this transformer. However, they only test the generated markup, that's why I focused the tests on the one thing the generated markup doesn't contain: the generated errors.

I can add a happy path though to ensure we touch most of the code even if there's something off with the spec tests...